activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1621,7 +1621,7 @@ def build_where_clause(opts, rest = []) # :nodoc:
             parts = [Arel.sql(opts)]
           elsif rest.first.is_a?(Hash) && /:\w+/.match?(opts)
             parts = [build_named_bound_sql_literal(opts, rest.first)]
-          elsif opts.include?("?")
+          elsif opts.include?("?") && rest.compact.present?
```

Comment:
Thanks for the PR, but this isn't correct as it won't prepare fragment with some bound `nil` values, but:

  - That's not desirable, `nil` in bound values is fine
  - It doesn't fix the original issue if the nil bound value isn't alone, e.g. `.where("id = ? AND name IS ?", 42, nil)`